File: 1-500/0113. Path Sum II.java
// Medium
// DFS, backtracking
// Time:O(n^2), Space:O(n) 
// https://leetcode.cn/problems/path-sum-ii/

import java.util.*;

class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> path = new ArrayList<>();
        dfs(root, targetSum, path, res);
        return res;
    }

    private void dfs(TreeNode node, int target, List<Integer> path, List<List<Integer>> res) {
        if (node == null)
            return;
        
        path.add(node.val);
        target -= node.val;

        // check if reach leaf first and target = 0 -> collect to res
        if (node.left == null && node.right == null && target == 0) {
            res.add(new ArrayList<>(path));
        }

        dfs(node.left, target, path, res);
        dfs(node.right, target, path, res);

        path.remove(path.size() - 1);
    }
}

File: 0001-0500/0113. Path Sum II.py
# Time:O(n), Space:O(h)

from typing import List, Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        res = []

        def backtracking(node, path, cur_sum):
            if not node:
                return

            path.append(node.val)
            cur_sum += node.val

            if not node.left and not node.right:
                if cur_sum == targetSum:
                    res.append(path[:])

            backtracking(node.left, path, cur_sum)
            backtracking(node.right, path, cur_sum)

            path.pop()

        backtracking(root, [], 0)
        return res
   